internal/storage/image.go

```diff
@@ -398,6 +410,87 @@ func (svc *imageService) ImageStatus(systemContext *types.SystemContext, nameOrI
 	return &result, nil
 }
 
+// nolint:staticcheck
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	if systemContext.SignaturePolicyPath == "" {
+		log.Infof(ctx, "No policy applied, skipping check if image is allowed to run")
+		return nil
+	} else if userSpecifiedImage == "" {
+		return errors.New("client did not set the CRI Config.Image.UserSpecifiedImage field, please update it to use signature policies")
+	}
+
+	log.Debugf(ctx, "Getting image status of config image: %s", configImage)
+	configImageStatus, err := svc.ImageStatus(systemContext, configImage)
+	if err != nil {
+		return fmt.Errorf("image status from config image: %w", err)
+	}
+
+	// Check the cache first
+	var policyContext *signature.PolicyContext
+	if cacheItem := svc.policyContextCache.Get(systemContext.SignaturePolicyPath); cacheItem != nil {
+		log.Debugf(ctx, "Using policy context from cache")
+		policyContext = cacheItem.Value()
+	} else {
+		log.Debugf(ctx, "Creating policy context")
+		policy, err := signature.DefaultPolicy(systemContext)
+		if err != nil {
+			return fmt.Errorf("get default policy: %w", err)
+		}
+
+		policyContext, err = signature.NewPolicyContext(policy)
+		if err != nil {
+			return fmt.Errorf("create policy context: %w", err)
+		}
+
+		log.Debugf(ctx, "Caching policy context")
+		svc.policyContextCache.Set(systemContext.SignaturePolicyPath, policyContext, ttlcache.DefaultTTL)
+	}
+
+	log.Debugf(ctx, "Parsing docker reference from %s", userSpecifiedImage)
+	imageRef, err := docker.ParseReference("//" + userSpecifiedImage)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf(ctx, "Creating image source")
+	imageSrc, err := imageRef.NewImageSource(ctx, systemContext)
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			return fmt.Errorf("verifying policies against private registries it not supported yet: %w", err)
+		}
+		return fmt.Errorf("create image source: %w", err)
+	}
+
+	// Connect the on-registry image with the already locally-pulled one
+	imageCloser, err := svc.PrepareImage(systemContext, userSpecifiedImage)
+	if err != nil {
+		return fmt.Errorf("prepare user specified image: %w", err)
+	}
+	defer imageCloser.Close()
+	if imageCloser.ConfigInfo().Digest.Encoded() != configImageStatus.ID {
+		return errors.New("local and remote image identifiers do not match")
+	}
+
+	log.Debugf(ctx, "Check if image %s is allowed to run", userSpecifiedImage)
+	unparsedImage := cimage.UnparsedInstance(imageSrc, &configImageStatus.Digest)
```

Comment:
Yes, for multi-arch images this must choose the right instance … but `configImageStatus.Digest` could be a value from _any_ of the deduplicated images; it might not be a digest that exists on this remote registry/repo (or it might exist but it might not be signed with `userSpecifiedImage`.

The right one can be determined by
- if the input reference is a repo@digest one, doing a storage lookup on that repo@digest
- If the input reference is an image ID, doing a storage lookup on userSpecifiedImage@imageID (which might be repo@digest@imageID)

… and, in both cases, the @digest value might refer to a manifest list; in that case it needs to be turned into the right per-arch instance if `manifest.MIMETypeIsMultiImage`, via `manifest.ListFromBlob` and the like. (I vaguely remember we’ve run into this lookup needing an enhancement when dealing with multiple architectures, but I can’t now find that CRI-O issue.)

Reply:
This code path does not exist any more.